timemachine/cpp/src/kernels/k_nonbonded.cuh

```diff
@@ -74,6 +74,41 @@ void __global__ k_coords_to_kv(
 
 }
 
+template <typename RealType>
+void __global__ k_check_rebuild(
+    const int N,
+    const double *new_coords,
+    const double *old_coords,
+    double padding,
+    int *rebuild) {
+
+    const int atom_idx = blockIdx.x*blockDim.x + threadIdx.x;
+
+    if(atom_idx >= N) {
+        return;
+    }
+
+    RealType xi = old_coords[atom_idx*3+0];
+    RealType yi = old_coords[atom_idx*3+1];
+    RealType zi = old_coords[atom_idx*3+2];
+
+    RealType xj = new_coords[atom_idx*3+0];
+    RealType yj = new_coords[atom_idx*3+1];
+    RealType zj = new_coords[atom_idx*3+2];
+
+    RealType dx = xi - xj;
+    RealType dy = yi - yj;
+    RealType dz = zi - zj;
+
+    RealType d2ij = dx*dx + dy*dy + dz*dz;
+    if(d2ij > static_cast<RealType>(0.25)*padding*padding) {
+        // (ytz): this is *safe* but technically is a race condition
+        rebuild[0] = 1;
+    }
+
+
+}
```

Comment:
Should this function also check if the box has changed? (Or does changing the box automatically trigger a neighborlist rebuild elsewhere in the code)?

(Consider a chain of neighboring particles x[i] in the center of a very large box. If I shrink the box so that the borders are close to x[0] and x[N], then (x[0], x[N]) could now be neighbors, even though none of the particle coordinates x[i] changed.)

Reply:
Hm. Currently we're running only in NVT, with NPT this definitely would be a problem, and we'd need to rebuild the neighborlist if the box size changes.

Reply:
Even without running an NPT simulation, I guess we could also run into an issue if the neighbor list was last computed with some box size, and we later query energies/forces/... for a different, smaller box size. (For example, if we also allow to optimize the box vectors during energy minimization.)

Perhaps we should add a check that `old_box` and `new_box` are the same before calling `k_check_rebuild`?

Reply:
This seems like a reasonable thing to add. I'll fix.